File: src/main/kotlin/com/github/vitormbgoncalves/kotlincourse/interop/CustomerRepository.java
package com.github.vitormbgoncalves.kotlincourse.interop;

import com.github.vitormbgoncalves.kotlincourse.classes.CustomerJava;

import java.util.List;

/**
 *<p>
 *  Sample Java Interface
 *</p>
 *<p>
 * @author Vitor Goncalves
 * @since 08.04.2021, qui, 18:54
 *</p>
 */

public interface CustomerRepository {
    CustomerJava getById(int id);

    List<CustomerJava> getAll();
}

File: src/main/kotlin/com/github/vitormbgoncalves/kotlincourse/inheritance/CustomerRepository.kt
package com.github.vitormbgoncalves.kotlincourse.inheritance

/**
 * Interfaces in Kotlin
 *
 * @author Vitor Goncalves
 * @since 06.04.2021, ter, 10:08
 */

interface CustomerRepository {
    val isEmpty: Boolean
        get() = true

    fun store(obj: Customer) {
        // implement code to store
    }

    fun getById(id: String): Customer
}

class SQLCustomerRepository : CustomerRepository {
    override fun getById(id: String): Customer {
        TODO("Not yet implemented")
    }

    override val isEmpty: Boolean
        get() = false
}

interface Interface1 {
    fun funA() {
        println("Fun A from Interfaces 1")
    }
}

interface Interface2 {
    fun funA() {
        println("Fun A from Interfaces 2")
    }
}

class Class1And2 : Interface1, Interface2 {
    override fun funA() {
        super<Interface2>.funA()
    }
}

fun main(args: Array<String>) {
    val c = Class1And2()
    c.funA()
}
